### ISSUE TITLE
Post scheduling

### ISSUE THREAD

**Issue by ludwiczakpawel** [forem/forem#15858]
### TL;DR

Let users schedule their post's publication date and time. 

### Description

Add a UI control (ideally [`input[type="datetime-local"]`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/datetime-local) as it comes with browser's native handy controls for date and time picker) to let authors schedule their posts instead of immediate publishing. The [flow is visualized in the Figma mockup](https://www.figma.com/file/zIGNwz3ahbZY7o8rlq1JeL/Post-scheduling?node-id=0%3A1).

Additionally, this will require also tiny changes elsewhere:
- Article view: once you schedule a post you should be able to see it as well, similarly to what happens when you save it as a draft. Let's [utilize that message](https://user-images.githubusercontent.com/108287/147069175-e0ff908d-9fe8-49ae-ac4f-fdf619f3a3e0.png) on top of the post to show appropriate message, e.g. "_This post publication has been scheduled for mm-dd-yyyy, hh:mm_" or something like that :).
- Dashboard: we may need to [add extra badge](https://user-images.githubusercontent.com/108287/147069508-cba79caa-6d2f-40ff-a0a8-82978b0b8f83.png) there to indicate that a post has been scheduled.

### Resources 

- [Related Slack Thread](https://forem-team.slack.com/archives/C7GE84DEJ/p1640037648469500)
- [Mockups](https://www.figma.com/file/zIGNwz3ahbZY7o8rlq1JeL/Post-scheduling?node-id=0%3A1)

**Comment by JennieOcken** [comment]
FYI: https://github.com/forem/forem/discussions/15473

**Comment by lightalloy** [comment]
I described my ideas in a [forem team post](https://forem.team/anna/scheduling-posts-implementation-527k), and we had a productive discussion in the comments section.
Let's break down the feature:

## Make it possible to set `published_at` field to the future

- [ ] change the validation
#### Setting `published_at` from editor v1
- [ ]  Frontend: setting the default published_at similar to other fields, and passing it when submitting a form. I suppose that the default value may be commented similar to the `cover_image`.
- [ ]  Backend: parse from frontmatter, make sure that `published_at` is set
#### Setting `published_at` from the rich markdown editor
- [ ]  add a datetime picker to the post options
- [ ]  make sure that the date from the datetime picker is passed correctly and saved
- [ ]  don't allow to edit `published_at` in the post options when the post has been published already
- [ ]  change the button label from “Publish now” to “Schedule” once the schedule publication date is set. Additionally, show tooltip on hover.

## Modify logic related to actions we do after the publication (notifications, mentions, slack notifications)
- [ ]  create a [notifications registry](https://forem.team/jeremy/comment/85m)
     ContextNotification model (`action` - name of the notification, `context_id`, `context_type` - article)
- [ ]  ? track sending slack notifications and creating mentions - I think that we can do without explicitly tracking creating mentions and sending slack notifications for now
- [ ]  create a worker that will:
   - find articles that were published (because `published_at` time has arrived) but no actions related to notifications were taken
   for those articles:
   - send slack notifications
   - create mentions
   - create notification subscriptions
   - create notifications
We have the code that performs these actions already, but we need to reorganize it: remove from `Articles::Creator` and `Article` callbacks.
- [ ] [Schedule the worker to run](https://forem.team/jamie/comment/860)

## Display scheduled articles 
- [ ] user dashboard: add status indicator and maybe little change in action buttons
- [ ] make sure that scheduled articles are not displayed in feeds (this should work already because filtering by `published_at` is already present in the `published` scope)
- [ ] article show (add a red notice similar to draft's one)
- [ ] make sure we render 404 when attempting to view a scheduled post as a non-author (?)
- [ ] add a state indicator on the articles page (`/admin/content_manager/articles`)
- [ ] change "Published at" to "Publish at" (`/admin/content_manager/articles`) for scheduled articles

## Misc
- [ ] setting featured (`/admin/content_manager/articles`) - I think we may keep the field as is and make sure that scheduled posts don't appear as featured
- [ ] setting approved (`/admin/content_manager/articles`) - I think we may keep the field as is and make sure that scheduled posts don't appear as approved in tags (they shouldn't already)
- [ ] pin post - we should make sure that we don't allow pinning scheduled posts, and hide the corresponding buttons on the admin page (/admin/content_manager/articles) and on users page (post => manage)

**Comment by lightalloy** [comment]
It seems like "quick fix" and "long term" options are not that different in terms of the time needed to implement, so I lean toward the "long-term" solution. Also, there is a large part of the feature that doesn't depend on what solution we choose.

**Comment by amywtlin** [comment]
I agree. Let's go with the long-term solution. Thank you @lightalloy!

**Comment by lightalloy** [comment]
The feature is merged, but remains behind a feature flag. The next steps are:
- Improve the datetime picker #18075 
- [Refactor saving `published_at` logic](https://github.com/forem/forem/pull/17939/files#r901483566)

**Comment by mirie** [comment]
@lightalloy can you remind me -- In your last comment there's a checkbox for [Refactor saving published_at logic](https://github.com/forem/forem/pull/17939/files#r901483566). Is that still a pending task?

**Comment by lightalloy** [comment]
@mirie yes, it's pending, sorry for the late response.
